### PR TITLE
Add rapids-doctor basic health checks

### PR DESCRIPTION
**Pull request by ncclementi**
## Description

Adding some basic health checks to be discovered by rapids-doctor when this is installed. 

Notes: 

- I want some feedback on weather the `functional_numba_check` is the right approach to test this, or if we should modify it. This was added based on my experience in certain scenarios I encountered, but this might not happen any more. Feedback welcome.
- I tested the usage and discoverability of the checks locally. 
- We can start with this checks, and continue with more as needed. 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes. (N/A)
- [ ] The documentation is up to date with these changes. (N/A)

cc: @jacobtomlinson for visibility

### CODE REVIEW

File: python/cudf/cudf/health_checks/_checks.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    df = cudf.DataFrame({"a": [1, 2, 2], "b": [10, 20, 30]})
+    res = df.groupby("a", as_index=False).agg({"b": "sum"})
+
+    if len(res) != 2:
+        raise AssertionError(
+            f"Expected 2 group rows after groupby/agg, got {len(res)}"
+        )
+
+    got_a = res["a"].to_pandas().tolist()
+    expected_a = [1, 2]
```
Comment (bdice) [inline python/cudf/cudf/health_checks/_checks.py]: Do we guarantee group ordering in cudf to be sorted and/or stable? Might want to check how other cudf tests are handling this.

File: python/cudf/cudf/health_checks/_checks.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    if got_a != expected_a:
+        raise AssertionError(
+            f"groupby key column mismatch: got {got_a}, expected {expected_a}"
+        )
+
+    got_b = res["b"].to_pandas().tolist()
```
Comment (bdice) [inline python/cudf/cudf/health_checks/_checks.py]: Rather than checking keys and values with separate assertions, should we just run cudf code and pandas code, then convert cuDF to Pandas and use `DataFrame.equals`?

File: python/cudf/cudf/_health_checks/_checks.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    Only the specific ``libnvvm.so`` missing-library ``OSError`` is rewritten
+    with install guidance; other ``OSError``s propagate unchanged.
+    """
+    import cudf
+
+    s = cudf.Series(["a", "aa", "b"])
+    try:
+        out = s.apply(lambda x: len(x))
+    except OSError as e:
+        if _is_libnvvm_missing_error(e):
```
Comment (bdice) [inline python/cudf/cudf/_health_checks/_checks.py:93]: Do you know how users get into this scenario? Is there a specific common mistake during installation?

File: python/cudf/cudf/health_checks/_checks.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
+#
+# SPDX-FileCopyrightText: Copyright (c) 2025-2026, NVIDIA CORPORATION.
+# SPDX-License-Identifier: Apache-2.0
+#
+
+"""cuDF health checks for rapids doctor."""
```
Comment (bdice) [inline python/cudf/cudf/health_checks/_checks.py]: This should probably link to https://github.com/rapidsai/rapids-cli. I had to dig for a while to find that repo, couldn't find it by searching "rapids doctor."

File: python/cudf/cudf/health_checks/_checks.py (hunk 1 above)
Comment (ncclementi) [reply python/cudf/cudf/health_checks/_checks.py]: Good catch, I'll make sure to modify this. I forgot that `Dataframe.equals` existed. I'll update this

File: python/cudf/cudf/_health_checks/_checks.py (hunk 3 above)
Comment (ncclementi) [reply python/cudf/cudf/_health_checks/_checks.py:93]: Yes, this usually happens when you are missing the cuda runtime. One common scenario is get a docker container like 

`docker run --gpus all -it nvidia/cuda:12.8.0-runtime-ubuntu24.04`

and then in this scenario, at least in the past (this is an older snippet I have saved), this had an issue:

```
# Then, in the container...
apt update && apt install python3 python3-pip
pip install --break-system-packages cudf-cu12
python3

# Then, this will work
>>> import cudf
>>> cudf.Series([1, 2, 3])

### but this not
cudf.Series([1, 2, 3]).apply(lambda x: len(x))
```

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/dist-packages/numba_cuda/numba/cuda/cudadrv/nvvm.py", line 139, in __new__
    inst.driver = open_cudalib('nvvm')
                  ^^^^^^^^^^^^^^^^^^^^

  File "/usr/local/lib/python3.12/dist-packages/numba_cuda/numba/cuda/cudadrv/libs.py", line 84, in open_cudalib
    return ctypes.CDLL(path)
           ^^^^^^^^^^^^^^^^^

  File "/usr/lib/python3.12/ctypes/__init__.py", line 379, in __init__
    self._handle = _dlopen(self._name, mode)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^
OSError: libnvvm.so: cannot open shared object file: No such file or directory
```

I'm not sure not sure if this will happen from now on, given cupy 14 and the pip wheels bringing all the ctk.

**Comment by ncclementi** [comment]
I updated the the code according to the review and answer the question regarding `_is_libnvvm_missing_error`. I'm noticing a few errors in CI, I rebase against main a few times but I still see failures but they seem unrelated to this PR.

File: python/cudf/cudf/health_checks/__init__.py
Comment (TomAugspurger) [inline python/cudf/cudf/health_checks/__init__.py:1]: Question for the cudf devs: do we want this to be public?

And for the rapids-cli devs: does this need to be public to be usable from `rapids-cli`? I guess it will show up in the `entrypoints`, so maybe public is for the best.

File: python/cudf/cudf/health_checks/_checks.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    # comparison by sorting on all result columns.
+    got = got.sort_values(by=list(got.columns)).reset_index(drop=True)
+    expected = res_pd.sort_values(by=list(res_pd.columns)).reset_index(
+        drop=True
+    )
+
+    if not got.equals(expected):
+        raise AssertionError(
+            "cuDF groupby/agg result does not match pandas after aligning order:\n"
+            f"cuDF (as pandas):\n{got}\n\npandas:\n{expected}"
```
Comment (TomAugspurger) [inline python/cudf/cudf/health_checks/_checks.py:51]: The other errors have additional guidance with an action the user can take (e.g. "follow the install guide"). Do we have any additional guidance here? Ask them to open an issue on github?

**Review by TomAugspurger — COMMENTED** [review]
A couple questions / comments about testing here:

1. It'd be good to ensure that at least the happy path is tested. Presumably `rapids-cli` will have some tests ensuring that things work from the CLI as expected. IMO, cudf should verify that at least the happy path is tested (by calling `functional_check()` and just making sure there's no exception).
2. And for the not happy paths... these seem hard to test. Do we try to jump through hoops with mocks and stuff to trigger the errors? IMO it's not worth it.

File: python/cudf/cudf/health_checks/__init__.py
Comment (bdice) [reply python/cudf/cudf/health_checks/__init__.py:1]: I would prefer for this not to be exposed publicly in the `cudf` namespace. We try to keep that tidy and aligned with pandas to some extent.

**Review by bdice — APPROVED** [review]
This seems fine, though @TomAugspurger raised a good point that we probably don't want this to be a public submodule.

**Comment by bdice** [comment]
Also please verify that the health checks run in our _own_ cuDF CI jobs. I don't want the library to ship with failing health checks due to code changes.

File: python/cudf/cudf/health_checks/__init__.py
Comment (ncclementi) [reply python/cudf/cudf/health_checks/__init__.py:1]: IIUC we need this to be public for the entry point to pick it up, but maybe I'm missing something. @jacobtomlinson Would you mind chiming in here?

File: python/cudf/cudf/health_checks/_checks.py (hunk 5 above)
Comment (ncclementi) [reply python/cudf/cudf/health_checks/_checks.py:51]: That's a good point, we can point them to open an issue on github.

**Comment by ncclementi** [comment]
> * It'd be good to ensure that at least the happy path is tested. Presumably `rapids-cli` will have some tests ensuring that things work from the CLI as expected. IMO, cudf should verify that at least the happy path is tested (by calling `functional_check()` and just making sure there's no exception).
> * And for the not happy paths... these seem hard to test. Do we try to jump through hoops with mocks and stuff to trigger the errors? IMO it's not worth it.

To answer some of the questions, we have some cli test that ensure discoverability in general. BUt we don't have specific ones for cudf. 

When cuml added checks, they implemented some form of discoverability tests themselves because they also added a separate health check system besides rapids-doctor, see https://github.com/rapidsai/cuml/pull/7854/changes#diff-f6c8b0321c6954f50907077aaf0ebbf56d4ca0d4a1476070c5c44731cda42858. But when I chatted with @jacobtomlinson about this, he mentioned these were no necessary, but I'll let him chime in here. 

> Also please verify that the health checks run in our own cuDF CI jobs. I don't want the library to ship with failing health checks due to code changes.

@bdice I'm not sure I follow. Do you want us to add a job in CI or add it as part of an existing one, and had rapids-cli as a dependency and run this? 

I was following the process that cuml had, and they are not running this in CI, but maybe we should all discuss how do we want these checks to be added.

@jacobtomlinson maybe you want to chime in here to add more context, since you discuss this at some point with the teams

**Comment by bdice** [comment]
> Do you want us to add a job in CI or add it as part of an existing one, and had rapids-cli as a dependency and run this?

Either way. Just something that ensures that the code added in this PR is executed, and that health checks pass in a known-good environment.

File: python/cudf/cudf/_health_checks/__init__.py
Comment (jacobtomlinson) [reply python/cudf/cudf/_health_checks/__init__.py:1]: I _think_ it can be private. It just needs to be referenced in the `pyproject.toml` correctly. It's not like anything in Python is actually truly private so I would expect entrypoints like `cudf._health_checks.foo` would still work.

File: python/cudf/cudf/_health_checks/_checks.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    # comparison by sorting on all result columns.
+    got = got.sort_values(by=list(got.columns)).reset_index(drop=True)
+    expected = res_pd.sort_values(by=list(res_pd.columns)).reset_index(
+        drop=True
+    )
+
+    if not got.equals(expected):
+        raise AssertionError(
+            "cuDF groupby/agg result does not match pandas after aligning order:\n"
+            f"cuDF (as pandas):\n{got}\n\npandas:\n{expected}"
```
Comment (jacobtomlinson) [reply python/cudf/cudf/_health_checks/_checks.py:51]: One thing I want to try and enforce is that every error message has some kind of actionable advice. Even if that is just a link to a troubleshooting guide. I don't think linking to opening an issue is a good idea. If we know enough about the problem to catch it in a check then we should be able to write something in the docs about resolving it and link to that.

**Comment by jacobtomlinson** [comment]
We should definitely test the happy path in CI. Mostly just to ensure the checks are actually importable/callable. I think it is ok not to test the unhappy paths, unless it looks easy to do. Cramming in a load of mocks to reproduce some edge case environment error reduces maintainability in a way that would outweigh the benefits of doing this. To parrot what @mmccarty was recommending, we should use dependency injection where possible when writing checks because it makes it way easier to test unhappy paths without using mocks. 

> Also please verify that the health checks run in our own cuDF CI jobs. I don't want the library to ship with failing health checks due to code changes.

The simplest way is to just add some pytest tests for the checks that get run as part of the normal test suite. The two main things you want to test are whether the checks are registered as entrypoints, conform to the expected signature and that it returns cleanly on the happy path. https://github.com/rapidsai/rapids-cli#check-plugins

We could include some type information in rapids-cli and check the signature that way, or add some testing helpers that could be reused across many libraries. But that would make rapids-cli a test time dependency of cudf. I think that's fine, but @bdice do you have any thoughts on that? 

It would also be nice to run a simple end to end test like calling `rapids doctor cudf --verbose` in CI and then assert that it exits cleanly and that all the entrypoints listed in `pyproject.toml` are present in the output. 

> [cuml] also added a separate health check system besides rapids-doctor

I strongly want to discourage this and wish I could've talked them out of doing this. If every library implements their own CLI machinery for calling the checks it undermines the user story of `rapids doctor` being the consistent place you go to debug your environment. I especially want to encourage that projects include "Paste the output of `pip install rapids-cli && rapids doctor`" in their issue template, and not something like "Paste the output of `python -m cudf.run_doctor_checks`". I understand the desire to keep debug dependencies to a minimum, but I think the consistency across the libraries outweighs that. The `rapids-cli` package is designed to be light.

**Comment by bdice** [comment]
> But that would make rapids-cli a test time dependency of cudf. I think that's fine, but @bdice do you have any thoughts on that?

As long as we avoid circular dependencies and `rapids-cli` folks help maintain these health checks if they break for non-cudf reasons, that's fine.

**Comment by ncclementi** [comment]
I'm  seeing a failure in the [conda-python-cudf-tests / 12.2.2, 3.11, amd64, rockylinux8, l4, earliest-driver, oldest-deps](https://github.com/NVIDIA/cudf/actions/runs/33121224812/job/98694571694?pr=21928#logs) case, I just modify the code to have a bit more information, but it seems something very weird is happening because 

I get this, which seems a segfault, although the checks passed. 

```bash
  File "/__w/cudf/cudf/python/cudf/cudf/tests/test_health_checks.py", line 39, in test_rapids_doctor_runs_required_and_cudf_health_checks
    assert result.returncode == 0, output
AssertionError: 🧑‍⚕️ Performing REQUIRED health check for RAPIDS 
  Running checks
  All checks passed!
  
  
assert -11 == 0
 +  where -11 = CompletedProcess(args=['rapids', 'doctor'], returncode=-11, stdout='🧑\u200d⚕️ Performing REQUIRED health check for RAPIDS \nRunning checks\nAll checks passed!\n', stderr='').returncode
```

Waiting for the current run to finish to see if I can get more info on why this is failing. That being said I'd need a bit of help here since this seems to only be failing on oldest dependencies path. 

cc: @bdice

File: python/cudf/cudf/tests/test_health_checks.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        timeout=120,
+    )
+
+    # A successful RAPIDS Doctor exit code means all discovered checks passed,
+    # including the registered cuDF checks validated above.
+    assert result.returncode == 0, (
+        f"rapids doctor exited with return code {result.returncode}\n"
+        f"stdout:\n{result.stdout}\n"
+        f"stderr:\n{result.stderr}"
+    )
```
[the remark this answers is not in the document]
Comment (ncclementi) [reply python/cudf/cudf/tests/test_health_checks.py:42]: resolved in [6a34de5](https://github.com/NVIDIA/cudf/pull/21928/commits/6a34de5a8dd2dd3d8fc39d50f6ef14c3df088b0e)

**Comment by ncclementi** [comment]
So I'm not sure what's happening, it seems that some how the checks passed but it seems like I'm hitting a segfault (that -11 I believe means segfault here) when comparing the return codes. It only happens for  the case of old dependecies, see https://github.com/NVIDIA/cudf/actions/runs/33404951587/job/99539765205?pr=21928#step:12:3585 

Any idea what could be happening here? 

cc: @bdice for when you can take a look.

**Comment by bdice** [comment]
@ncclementi This is a bug in `cuda.core` teardown. I filed https://github.com/NVIDIA/cuda-python/issues/2743. However, the easiest fix is to bump our lower bounds on `cuda-bindings` from `>=12.9.2` to `>=12.9.3` for CUDA 12, and `>=13.0.1` to `>=13.0.2` for CUDA 13. I pushed that in 166e8f7ed5.

**Review by bdice — APPROVED** [review]

**Comment by ncclementi** [comment]
Just checking, is there anything outstanding here, or can this be merged? I don't think I have merge powers here so I'd need someone to do it. 
cc: @bdice @vyasr
